gh-91048: Chain some exceptions in _testexternalinspection.c

sergey-miryanov opened the pull request:
1. Added extra Py_UNREACHABLE calls on unsupported platforms
2. Remove unused refcnt read
3. Set result to 0 if fd closing failed
4. Chain some exceptions to get more info if failed

<!-- gh-issue-number: gh-91048 -->
* Issue: gh-91048
<!-- /gh-issue-number -->

sergey-miryanov committed:
Add Py_UNREACHABLE

sergey-miryanov committed:
Remove unused refcnt

sergey-miryanov committed:
Reset result if can not close file descriptor

sergey-miryanov committed:
Set exception only if not set

sergey-miryanov committed:
Chain exception to get more info

picnixz commented on Modules/_testexternalinspection.c:
```diff
@@ -23,6 +23,18 @@
+#error "CHAIN_EXCEPTIONS should not be defined"
+#endif
+
+#define CHAIN_EXCEPTIONS(Type, Msg)                 \
```
Nit but in general, if it's not speed-critical, and when it's possible, we prefer static functions over macros (in this case, a static function is fine and it helps typing)

sergey-miryanov replied:
Fixed, thanks!

sergey-miryanov committed:
Remove error string from macos impl to keep PermissionError

sergey-miryanov committed:
Replace macro with static function

sergey-miryanov committed:
Remove chain_exceptions from remote_debug to make clang on macos happy

- I believe it is not worth to add conditional compilation of
  such small function like chain_exceptions

pablogsal reviewed (approved):
LGTM

This is a great improvement. Thanks a lot!
